### ISSUE TITLE
Reconnecting with the engine

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#5202]
What IDE should do:
1. Once we lose the connection with LS, try to reconnect.
2. When reconnected, is should reinitialize the protocol, restore all execution contexts and visualizations, and synchronize the code

This task is automatically imported from the old Task Issue Board and it was originally created by Adam Obuchowicz.
Original issue is [here](https://www.pivotaltracker.com/story/show/184372062).
<hr />

### Why

**As Enso IDE user**
**I want ** to IDE try to reconnect to the Language Server if the connection is lost
**So that ** I can continue working once the LS is restored.

### Discussion
https://discord.com/channels/401396655599124480/1050713090829852712

### Acceptance Criteria

```gherkin
Scenario: 
Given IDE with already open project
When language server is closed or otherwise connection is lost
Then IDE should attempt to reconnect to the language server and reinitialize the session
```

**Notes:**

```[tasklist]
### GUI Tasks
- [ ] https://github.com/enso-org/enso/issues/8520
- [ ] https://github.com/enso-org/enso/issues/8522
- [ ] https://github.com/enso-org/enso/issues/8523
```

**Comment by Frizi** [comment]
TL;DR from discord:

step by step reproduction scenario by @hubertp:
> 1. start gui
> 2. start backend (or rather project-manager) separately
> 3. enter any project and let it boot
> 4. kill backend (that includes project-manager and LS of the project)
> 5. let IDE cook for a while until it notices the problem and provides a warning 
> 6. start backend again

ide side task description by @farmaazon:
> on the IDE side, we would need to do the following:
> 1. once the we lose connection with LS, try to reconnect. 
> 2. When reconnected, we should reinitialise the protocol, restore all execution contexts and visualisations, and synchronise the code

**Comment by hubertp** [comment]
Doc on the current state of reconnections https://docs.google.com/document/d/1yah7vK9NDt9l15XOXNak37hMugPFmxUsoezrnsyUEU4/edit?usp=sharing

<blockquote><div><strong><a href="https://docs.google.com/document/d/1yah7vK9NDt9l15XOXNak37hMugPFmxUsoezrnsyUEU4/edit?usp=sharing">Google Docs: Sign-in</a></strong></div><div>Access Google Docs with a personal Google account or Google Workspace account (for business use).</div></blockquote>

**Comment by jdunkerley** [comment]
@hubertp to review state against GUI2.
Need to ensure Windows Hibernation/Standby support works.

**Comment by farmaazon** [comment]
As I know, GUI2 does not take any action to reconnect with Language Server. Created subtasks to implement it.
